lib/app/server.ts

```diff
@@ -0,0 +1,437 @@
+        const urlOptionsAllowed = ['readOnly', 'hideEditorToolbars', 'language'];
+        const filteredUrlOptions = _.mapObject(_.pick(urlOptions || {}, urlOptionsAllowed), val =>
+            utils.toProperty(val),
+        );
+        const allExtraOptions = _.extend({}, filteredUrlOptions, extra);
+
+        if (allExtraOptions.mobileViewer && allExtraOptions.config) {
+            const clnormalizer = new normalizer.ClientStateNormalizer();
+            clnormalizer.fromGoldenLayout(allExtraOptions.config);
+            const clientstate = clnormalizer.normalized;
+
+            const glnormalizer = new normalizer.ClientStateGoldenifier();
+            allExtraOptions.slides = glnormalizer.generatePresentationModeMobileViewerSlides(clientstate);
+        }
+
+        const options = _.extend({}, allExtraOptions, clientOptionsHandler.get());
+        options.optionsHash = clientOptionsHandler.getHash();
+        options.compilerExplorerOptions = JSON.stringify(allExtraOptions);
+        options.extraBodyClass = options.embedded ? 'embedded' : extraBodyClass;
+        options.httpRoot = httpRoot;
+        options.staticRoot = staticRoot;
+        options.storageSolution = 'local'; // TODO: Don't hardcode this
+        options.require = pugRequireHandler;
+        options.sponsors = sponsorConfig;
+        return options;
+    }
+
+    /**
+     * Renders GoldenLayout for a given configuration
+     */
+    function renderGoldenLayout(
+        config: GoldenLayoutRootStruct,
+        metadata: ShortLinkMetaData,
+        req: Request,
+        res: Response,
+    ) {
+        const embedded = req.query.embedded === 'true';
+
+        res.render(
+            embedded ? 'embed' : 'index',
+            renderConfig(
+                {
+                    embedded: embedded,
+                    mobileViewer: isMobileViewer(req),
+                    config: config,
+                    metadata: metadata,
+                    storedStateId: req.params.id || false,
+                },
+                req.query,
+            ),
+        );
+    }
+
+    /**
+     * Handles rendering embedded pages
+     */
+    const embeddedHandler = (req: Request, res: Response) => {
+        res.render(
+            'embed',
+            renderConfig(
+                {
+                    embedded: true,
+                    mobileViewer: isMobileViewer(req),
+                },
+                req.query,
+            ),
+        );
+    };
+
+    /**
+     * Handles legacy Google URL shortener redirects
+     */
+    function oldGoogleUrlHandler(req: Request, res: Response, next: NextFunction) {
+        const id = req.params.id;
+        const googleUrl = `https://goo.gl/${encodeURIComponent(id)}`;
+        googleShortUrlResolver
+            .resolve(googleUrl)
+            .then(resultObj => {
+                const parsed = new url.URL(resultObj.longUrl);
+                const allowedRe = new RegExp(ceProps<string>('allowedShortUrlHostRe'));
```

Comment:
Same here, and change to use async/await

Reply:
Re: "Same here" -> not clear what the "same" is
Re: async/await -> done

Reply:
Ah, move the functions to the top-level, out of the current function